README.md
# photo_selector
一个Flutter轻量级的照片选择器,支持iOS和Android.
## 功能介绍
* 类似微信朋友圈九宫格照片预览器
* 可编辑的照片选择器
* 可设置水平间距、垂直间距
* 可设置列数
* 可设置是否可滚动
* 可设置是否可编辑
* 可设置最大数量

## 添加依赖

```yaml
dependencies:
  photo_selector: ^0.0.1
```
## 照片控制器
```dart
/// 照片选择器
class PhotoSelector extends StatefulWidget {
  /// 照片数据集合
  List<PhotoData> _photoDatas = [];

  /// 照片选择器布局
  final PhotoSelectorLayout _photoSelectorLayout = PhotoSelectorLayout();

  /// 照片选择器回调处理
  final PhotoSelectorHandler _photoSelectorHandler = PhotoSelectorHandler();

  /// 创建状态
  @override
  State<StatefulWidget> createState() => _PhotoSelectorState();

  /// 构造照片选择器
  PhotoSelector({
    required List<PhotoData> photoDatas,
    void Function(PhotoSelectorLayout photoSelectorLayout)? photoSelectorLayout,
    void Function(PhotoSelectorHandler photoSelectorHandler)?
        photoSelectorHandler,
  }) {
    _photoDatas = photoDatas;
    photoSelectorLayout?.call(_photoSelectorLayout);
    photoSelectorHandler?.call(_photoSelectorHandler);
  }
}
```
## 布局类
```dart
/// 照片选择器布局
class PhotoSelectorLayout extends Object {
  /// 列数
  int column;

  /// 最大数量
  int? maxCount;

  /// 垂直间距
  double verticalSpacing;

  /// 水平间距
  double horizontalSpacing;

  /// 编辑
  bool edit;

  /// 滚动
  bool scroll;

  /// 构建照片选择器布局
  PhotoSelectorLayout({
    this.column = 3,
    this.horizontalSpacing = 5.0,
    this.verticalSpacing = 5.0,
    this.edit = true,
    this.scroll = true,
  });
}
```
## 处理类
```dart
/// 照片选择器处理
class PhotoSelectorHandler extends Object {
  /// 预览回调
  void Function(PhotoData photoData, List<PhotoData> photoDatas)? onPreview;

  /// 删除回调
  void Function(PhotoData photoData, List<PhotoData> photoDatas)? onDelete;

  /// 添加回调
  void Function(void Function(List<PhotoData> photoDatas))? onAdd;

  /// 照片选择器回调处理
  PhotoSelectorHandler({this.onAdd, this.onPreview, this.onDelete});
}
```
## 示例
```dart
import 'dart:io';
import 'package:flutter/material.dart';
import 'package:image_picker/image_picker.dart';
import 'package:photo_selector/photo_selector.dart';

void main() {
  runApp(MaterialApp(
    home: MyApp(),
  ));
}
class MyApp extends StatelessWidget {
  List<PhotoData> photos = [
    PhotoData(
      allowEdit: false,
      imageData:
          "https://gimg2.baidu.com/image_search/src=http%3A%2F%2Fbpic.588ku.com%2Felement_origin_min_pic%2F18%2F08%2F24%2F05dbcc82c8d3bd356e57436be0922357.jpg&refer=http%3A%2F%2Fbpic.588ku.com&app=2002&size=f9999,10000&q=a80&n=0&g=0n&fmt=jpeg?sec=1637479550&t=8c5f21b7a3e516e74eba3b72fb63ecce",
    ),
    PhotoData(
      imageData:
          "https://gimg2.baidu.com/image_search/src=http%3A%2F%2Fimage66.360doc.com%2FDownloadImg%2F2013%2F11%2F0708%2F36512556_53.jpg&refer=http%3A%2F%2Fimage66.360doc.com&app=2002&size=f9999,10000&q=a80&n=0&g=0n&fmt=jpeg?sec=1637563329&t=cd4c36d777fa2d5cc3c60ce866a22a9a",
    ),
    PhotoData(
      imageData:
          "https://gimg2.baidu.com/image_search/src=http%3A%2F%2Fpic1.nipic.com%2F2009-02-26%2F2009226195511805_2.jpg&refer=http%3A%2F%2Fpic1.nipic.com&app=2002&size=f9999,10000&q=a80&n=0&g=0n&fmt=jpeg?sec=1637563329&t=f21cc4dc3226c429adf45909367080fb",
    ),
    PhotoData(
      imageData:
          "https://gimg2.baidu.com/image_search/src=http%3A%2F%2Fpic14.nipic.com%2F20110512%2F6202394_223408457128_2.jpg&refer=http%3A%2F%2Fpic14.nipic.com&app=2002&size=f9999,10000&q=a80&n=0&g=0n&fmt=jpeg?sec=1637563329&t=91c41dd3c8338e63a489624fa35b7e3e",
    ),
    PhotoData(
      imageData:
          "https://gimg2.baidu.com/image_search/src=http%3A%2F%2Fimg.daimg.com%2Fuploads%2Fallimg%2F140628%2F3-14062Q52P5.jpg&refer=http%3A%2F%2Fimg.daimg.com&app=2002&size=f9999,10000&q=a80&n=0&g=0n&fmt=jpeg?sec=1637563329&t=0f16a592da1207bb234640b533c83e84",
    ),
    PhotoData(
      imageData:
          "https://gimg2.baidu.com/image_search/src=http%3A%2F%2Fpic24.nipic.com%2F20121031%2F369125_100510566000_2.jpg&refer=http%3A%2F%2Fpic24.nipic.com&app=2002&size=f9999,10000&q=a80&n=0&g=0n&fmt=jpeg?sec=1637563329&t=7ed4c46b4b734829db15299d351b613b",
    ),
  ];

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: Text('照片选择器'),
      ),
      body: Container(
        padding: EdgeInsets.all(10),
        width: double.infinity,
        child: PhotoSelector(
          photoDatas: photos,
          photoSelectorLayout: (layout) {
            layout.edit = false;
            layout.maxCount = 9;
            layout.scroll = true;
            layout.horizontalSpacing = 5.0;
            layout.verticalSpacing = 5.0;
            layout.column = 3;
          },
          photoSelectorHandler: (handler) {
            handler.onAdd = (add) {
              ImagePicker().pickMultiImage().then((value) {
                List<PhotoData> photoDatas = [];
                value?.forEach((element) {
                  PhotoData photoData = PhotoData(
                    imageData: File(element.path),
                  );
                  photoDatas.add(photoData);
                });
                add(photoDatas);
                print("添加后的照片数量：${photos.length}");
              });
            };
            handler.onDelete = (data, datas) {
              print("删除的照片数据：${data.imageData}");
              print("删除后的照片数量：${datas.length}");
            };
            handler.onPreview = (data, datas) {
              print("当前预览的照片数据：${data.imageData}");
              print("当前预览的照片索引：${datas.indexOf(data)}");
              print("预览的照片数量：${datas.length}");
            };
          },
        ),
      ),
    );
  }
}
```
## 预览
![image](https://github.com/xtzPioneer/photo_selector/blob/main/project_preview/photo_selector_preview2.png?raw=true)
![image](https://github.com/xtzPioneer/photo_selector/blob/main/project_preview/photo_selector_preview3.png?raw=true)
![image](https://github.com/xtzPioneer/photo_selector/blob/main/project_preview/photo_selector_preview1.png?raw=true)


CHANGELOG.md
## 0.0.2

* 优化细节

test/photo_selector_test.dart
import 'package:flutter_test/flutter_test.dart';

import 'package:photo_selector/photo_selector.dart';

void main() {
  test('adds one to input values', () {
  });
}


lib/photo_selector.dart
library photo_selector;

import 'dart:io';
import 'package:flutter/material.dart';
import 'package:flutter/widgets.dart';

/// 包名
const String _packageName = "photo_selector";

/// 照片数据
class PhotoData extends Object {
  /// 照片数据
  var imageData;

  /// 允许编辑
  bool allowEdit;

  /// 构建照片数据
  PhotoData({
    required this.imageData,
    this.allowEdit = true,
  });
}

/// 照片选择器布局
class PhotoSelectorLayout extends Object {
  /// 列数
  int column;

  /// 最大数量
  int? maxCount;

  /// 垂直间距
  double verticalSpacing;

  /// 水平间距
  double horizontalSpacing;

  /// 编辑
  bool edit;

  /// 滚动
  bool scroll;

  /// 构建照片选择器布局
  PhotoSelectorLayout({
    this.column = 3,
    this.horizontalSpacing = 5.0,
    this.verticalSpacing = 5.0,
    this.edit = true,
    this.scroll = true,
  });
}

/// 照片选择器处理
class PhotoSelectorHandler extends Object {
  /// 预览回调
  void Function(PhotoData photoData, List<PhotoData> photoDatas)? onPreview;

  /// 删除回调
  void Function(PhotoData photoData, List<PhotoData> photoDatas)? onDelete;

  /// 添加回调
  void Function(void Function(List<PhotoData> photoDatas))? onAdd;

  /// 照片选择器回调处理
  PhotoSelectorHandler({this.onAdd, this.onPreview, this.onDelete});
}

/// 照片选择器
class PhotoSelector extends StatefulWidget {
  /// 照片数据集合
  List<PhotoData> _photoDatas = [];

  /// 照片选择器布局
  final PhotoSelectorLayout _photoSelectorLayout = PhotoSelectorLayout();

  /// 照片选择器回调处理
  final PhotoSelectorHandler _photoSelectorHandler = PhotoSelectorHandler();

  /// 创建状态
  @override
  State<StatefulWidget> createState() => _PhotoSelectorState();

  /// 构造照片选择器
  PhotoSelector({
    required List<PhotoData> photoDatas,
    void Function(PhotoSelectorLayout photoSelectorLayout)? photoSelectorLayout,
    void Function(PhotoSelectorHandler photoSelectorHandler)?
        photoSelectorHandler,
  }) {
    _photoDatas = photoDatas;
    photoSelectorLayout?.call(_photoSelectorLayout);
    photoSelectorHandler?.call(_photoSelectorHandler);
  }
}

/// 照片选择器状态
class _PhotoSelectorState extends State<PhotoSelector> {
  /// Build
  @override
  Widget build(BuildContext context) {
    return GridView.count(
      crossAxisSpacing: widget._photoSelectorLayout.horizontalSpacing,
      mainAxisSpacing: widget._photoSelectorLayout.verticalSpacing,
      crossAxisCount: widget._photoSelectorLayout.column,
      physics: widget._photoSelectorLayout.scroll
          ? null
          : NeverScrollableScrollPhysics(),
      children: _cells(),
    );
  }

  /// 单元格集合
  List<Widget> _cells() {
    // 单元格集合
    List<Widget> cells = [];

    // 添加PhotoPreviewCell
    widget._photoDatas.forEach((element) {
      _PhotoPreviewData photoPreviewData = _PhotoPreviewData(
        imageData: element.imageData,
        edit: widget._photoSelectorLayout.edit ? element.allowEdit : false,
        onPreview: (data) {
          _preview(element);
        },
        onDelete: (data) {
          widget._photoDatas.remove(element);
          _delete(element);
          setState(() {});
        },
      );
      _PhotoPreviewCell cell = _PhotoPreviewCell(photoPreviewData);
      cells.add(cell);
    });

    // 添加PhotoAddCell
    if (widget._photoSelectorLayout.maxCount != null) {
      int maxCount = widget._photoSelectorLayout.maxCount!;
      int currentCount = widget._photoDatas.length;
      if (maxCount > currentCount) {
        if (widget._photoSelectorLayout.edit) {
          cells.add(_PhotoAddCell(onAdd: _add));
        }
      } else {
        widget._photoDatas.removeRange(maxCount, currentCount);
        cells.removeRange(maxCount, currentCount);
      }
    } else {
      if (widget._photoSelectorLayout.edit) {
        cells.add(_PhotoAddCell(onAdd: _add));
      }
    }
    return cells;
  }

  /// 添加
  _add() {
    widget._photoSelectorHandler.onAdd?.call((photoDatas) {
      if (widget._photoSelectorLayout.maxCount != null) {
        int maxCount =
            widget._photoSelectorLayout.maxCount! - widget._photoDatas.length;
        int currentCount = photoDatas.length;
        if (maxCount > currentCount) {
          widget._photoDatas.addAll(photoDatas);
        } else {
          photoDatas.removeRange(maxCount, photoDatas.length);
          widget._photoDatas.addAll(photoDatas);
        }
      } else {
        widget._photoDatas.addAll(photoDatas);
      }
      setState(() {});
    });
  }

  /// 预览
  _preview(var data) {
    widget._photoSelectorHandler.onPreview?.call(data, widget._photoDatas);
  }

  /// 删除
  _delete(var data) {
    widget._photoSelectorHandler.onDelete?.call(data, widget._photoDatas);
  }
}

/// 照片预览数据
class _PhotoPreviewData extends Object {
  /// 照片数据
  var imageData;

  /// 预览回调
  void Function(_PhotoPreviewData _photoPreviewData)? onPreview;

  /// 删除回调
  void Function(_PhotoPreviewData _photoPreviewData)? onDelete;

  /// 编辑
  bool edit;

  /// 构建照片预览数据
  _PhotoPreviewData({
    required this.imageData,
    this.onPreview,
    this.onDelete,
    this.edit = true,
  });
}

/// 照片预览单元格
class _PhotoPreviewCell extends StatelessWidget {
  /// 照片预览数据
  final _PhotoPreviewData _photoPreviewData;

  /// 构建照片预览单元格
  _PhotoPreviewCell(this._photoPreviewData);

  /// Build
  @override
  Widget build(BuildContext context) {
    return LayoutBuilder(builder: (context, constraints) {
      double deleteImageSize = constraints.maxWidth / 4.5;
      double borderRadius = constraints.maxWidth / 25.5;
      return Stack(
        children: [
          Positioned(
            child: GestureDetector(
              onTap: () {
                _photoPreviewData.onPreview?.call(_photoPreviewData);
              },
              child: Container(
                width: constraints.maxWidth,
                height: constraints.maxHeight,
                child: ClipRRect(
                  borderRadius: BorderRadius.circular(borderRadius),
                  child: Container(
                    color: Color.fromRGBO(244, 244, 244, 1),
                    child: _image(),
                  ),
                ),
              ),
            ),
          ),
          Positioned(
            right: 0,
            child: Offstage(
              offstage: !_photoPreviewData.edit,
              child: GestureDetector(
                onTap: () {
                  _photoPreviewData.onDelete?.call(_photoPreviewData);
                },
                child: Image.asset(
                  _ImageAsset().deleteImage,
                  package: _packageName,
                  fit: BoxFit.cover,
                  width: deleteImageSize,
                  height: deleteImageSize,
                ),
              ),
            ),
          ),
        ],
      );
    });
  }

  /// Image
  Image _image() {
    if (_photoPreviewData.imageData is String) {
      String string = _photoPreviewData.imageData;
      if (string.contains("http")) {
        return Image.network(
          _photoPreviewData.imageData,
          fit: BoxFit.cover,
        );
      } else {
        return Image.asset(
          _photoPreviewData.imageData,
          fit: BoxFit.cover,
        );
      }
    } else if (_photoPreviewData.imageData is File) {
      return Image.file(
        _photoPreviewData.imageData,
        fit: BoxFit.cover,
      );
    } else {
      return Image.asset(
        _ImageAsset().defaultImage,
        package: _packageName,
        fit: BoxFit.cover,
      );
    }
  }
}

/// 照片添加单元格
class _PhotoAddCell extends StatelessWidget {
  /// 添加回调
  final void Function()? onAdd;

  /// Build
  @override
  Widget build(BuildContext context) {
    return LayoutBuilder(
      builder: (context, constraints) {
        double addImageSize = constraints.maxWidth / 3.0;
        double addTextSize = constraints.maxWidth / 9.0;
        double borderRadius = constraints.maxWidth / 25.5;
        double addTextTop = constraints.maxWidth / 27.8;
        return GestureDetector(
          onTap: onAdd,
          child: Container(
            width: constraints.maxWidth,
            height: constraints.maxHeight,
            color: Colors.white,
            child: ClipRRect(
              borderRadius: BorderRadius.circular(borderRadius),
              child: Container(
                color: Color.fromRGBO(244, 244, 244, 1),
                child: Column(
                  mainAxisAlignment: MainAxisAlignment.center,
                  children: <Widget>[
                    Center(
                      child: Image.asset(
                        _ImageAsset().cameraImage,
                        package: _packageName,
                        width: addImageSize,
                        height: addImageSize,
                      ),
                    ),
                    Container(
                      padding: EdgeInsets.only(top: addTextTop),
                      child: Text(
                        '添加照片',
                        style: TextStyle(
                            fontSize: addTextSize, color: Color(0xff999999)),
                      ),
                    ),
                  ],
                ),
              ),
            ),
          ),
        );
      },
    );
  }

  /// 构建图片添加
  _PhotoAddCell({required this.onAdd});
}

/// Image资源
class _ImageAsset {
  final String cameraImage = "images/photo_selector_camera.png";
  final String defaultImage = "images/photo_selector_default.png";
  final String deleteImage = "images/photo_selector_delete.png";
}


example/lib/main.dart
import 'dart:io';
import 'package:flutter/material.dart';
import 'package:image_picker/image_picker.dart';
import 'package:photo_selector/photo_selector.dart';

void main() {
  runApp(MaterialApp(
    home: MyApp(),
  ));
}
class MyApp extends StatelessWidget {
  List<PhotoData> photos = [
    PhotoData(
      allowEdit: false,
      imageData:
      "https://gimg2.baidu.com/image_search/src=http%3A%2F%2Fbpic.588ku.com%2Felement_origin_min_pic%2F18%2F08%2F24%2F05dbcc82c8d3bd356e57436be0922357.jpg&refer=http%3A%2F%2Fbpic.588ku.com&app=2002&size=f9999,10000&q=a80&n=0&g=0n&fmt=jpeg?sec=1637479550&t=8c5f21b7a3e516e74eba3b72fb63ecce",
    ),
    PhotoData(
      imageData:
      "https://gimg2.baidu.com/image_search/src=http%3A%2F%2Fimage66.360doc.com%2FDownloadImg%2F2013%2F11%2F0708%2F36512556_53.jpg&refer=http%3A%2F%2Fimage66.360doc.com&app=2002&size=f9999,10000&q=a80&n=0&g=0n&fmt=jpeg?sec=1637563329&t=cd4c36d777fa2d5cc3c60ce866a22a9a",
    ),
    PhotoData(
      imageData:
      "https://gimg2.baidu.com/image_search/src=http%3A%2F%2Fpic1.nipic.com%2F2009-02-26%2F2009226195511805_2.jpg&refer=http%3A%2F%2Fpic1.nipic.com&app=2002&size=f9999,10000&q=a80&n=0&g=0n&fmt=jpeg?sec=1637563329&t=f21cc4dc3226c429adf45909367080fb",
    ),
    PhotoData(
      imageData:
      "https://gimg2.baidu.com/image_search/src=http%3A%2F%2Fpic14.nipic.com%2F20110512%2F6202394_223408457128_2.jpg&refer=http%3A%2F%2Fpic14.nipic.com&app=2002&size=f9999,10000&q=a80&n=0&g=0n&fmt=jpeg?sec=1637563329&t=91c41dd3c8338e63a489624fa35b7e3e",
    ),
    PhotoData(
      imageData:
      "https://gimg2.baidu.com/image_search/src=http%3A%2F%2Fimg.daimg.com%2Fuploads%2Fallimg%2F140628%2F3-14062Q52P5.jpg&refer=http%3A%2F%2Fimg.daimg.com&app=2002&size=f9999,10000&q=a80&n=0&g=0n&fmt=jpeg?sec=1637563329&t=0f16a592da1207bb234640b533c83e84",
    ),
    PhotoData(
      imageData:
      "https://gimg2.baidu.com/image_search/src=http%3A%2F%2Fpic24.nipic.com%2F20121031%2F369125_100510566000_2.jpg&refer=http%3A%2F%2Fpic24.nipic.com&app=2002&size=f9999,10000&q=a80&n=0&g=0n&fmt=jpeg?sec=1637563329&t=7ed4c46b4b734829db15299d351b613b",
    ),
  ];

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: Text('照片选择器'),
      ),
      body: Container(
        padding: EdgeInsets.all(10),
        width: double.infinity,
        child: PhotoSelector(
          photoDatas: photos,
          photoSelectorLayout: (layout) {
            layout.edit = true;
            layout.maxCount = 9;
            layout.scroll = true;
            layout.horizontalSpacing = 5.0;
            layout.verticalSpacing = 5.0;
            layout.column = 3;
          },
          photoSelectorHandler: (handler) {
            handler.onAdd = (add) {
              ImagePicker().pickMultiImage().then((value) {
                List<PhotoData> photoDatas = [];
                value?.forEach((element) {
                  PhotoData photoData = PhotoData(
                    imageData: File(element.path),
                  );
                  photoDatas.add(photoData);
                });
                add(photoDatas);
                print("添加后的照片数量：${photos.length}");
              });
            };
            handler.onDelete = (data, datas) {
              print("删除的照片数据：${data.imageData}");
              print("删除后的照片数量：${datas.length}");
            };
            handler.onPreview = (data, datas) {
              print("当前预览的照片数据：${data.imageData}");
              print("当前预览的照片索引：${datas.indexOf(data)}");
              print("预览的照片数量：${datas.length}");
            };
          },
        ),
      ),
    );
  }
}